File: leetcode_0449/cpp/leetcode_0449.cpp
//
// Created by wangzy on 2022/3/31.
//

#include<iostream>
#include<queue>
#include<vector>
#include <algorithm>
#include <unordered_map>
#include <list>
#include <deque>
#include <stack>
#include <sstream>
#include <typeinfo>
#include <climits>
#include <random>
#include <ctime>

using namespace std;

struct TreeNode{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0),left(nullptr),right(nullptr) {}
    TreeNode(int x) : val(x),left(nullptr),right(nullptr) {}
    TreeNode(int x,TreeNode *left,TreeNode* right):val(x),left(left),right(right) {}
};


class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        if(!root)
            return "#";
        return to_string(root->val)+" "+serialize(root->left)+" "+serialize(root->right);
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        stringstream ss(data);
        return rebuild(ss);
    }


    TreeNode* rebuild(stringstream& ss){
        string tmp;
        ss>>tmp;
        if(tmp == "#")
            return nullptr;
        TreeNode* node = new TreeNode(stoi(tmp));
        node->left = rebuild(ss);
        node->right = rebuild(ss);
        return node;
    }

    vector<string> levelOrderStr(TreeNode* root){
        vector<TreeNode*> cur;
        vector<TreeNode*> next;
        if(!root)
            return vector<string> {};
        if(root)
            cur.push_back(root);
        vector<string> res;
//        vector<string> curEles;
        while(cur.size()){
            //标识next是否都为null
            int flag = 0;
            for(int i = 0;i < cur.size();i++){
                TreeNode* node = cur[i];
                if(!node){
                    res.push_back("null");
                    continue;
                }
                else
                    res.push_back(to_string(node->val));
                //如果node的孩子不为null,则改变flag
                if(node->left || node->right)
                    flag++;
                next.push_back(node->left);
                next.push_back(node->right);
            }
            cur = next;
            if(flag == 0)
                cur.clear();
            next.clear();
        }
        return res;
    }

    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<TreeNode*> cur;
        vector<TreeNode*> next;
        if(!root)
            return vector<vector<int>> {};
        if(root)
            cur.push_back(root);
        vector<vector<int>> res;
        vector<int> curEles;
        while(cur.size()){
            for(int i = 0;i < cur.size();i++){
                TreeNode* node = cur[i];
                curEles.push_back(node->val);
                if(node->left)
                    next.push_back(node->left);
                if(node->right)
                    next.push_back(node->right);
            }
            res.push_back(curEles);
            curEles.clear();
            cur = next;
            next.clear();
        }
        return res;
    }


    void traversePreTree(string& s,TreeNode* root){
        if(root){
            s+=(to_string(root->val)+",");
            traversePreTree(s,root->left);
            traversePreTree(s,root->right);
        }
    }
    vector<string> printTree(TreeNode* root){
        string s;
        vector<string> levelOrder = levelOrderStr(root);
        return levelOrder;
    }
    void deleteTree(TreeNode* root){
        if(root){
            deleteTree(root->left);
            deleteTree(root->right);
            root = nullptr;
        }
    }
};

template<typename T>
void printVec(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

template<typename T>
string convertVecToStr(std::vector<T> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << ele;
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    return res;
}

template<typename T>
void printVecArr(std::vector<std::vector<T>> l){
    std::string res = "[";
    std::string tmp = "";
    std::stringstream ss;
    for(auto ele : l){
        ss << convertVecToStr(ele);
        ss >> tmp;
        res += tmp + ",";
        ss.clear();
    }
    res = res.substr(0,res.size()-1);
    res += "]";
    std::cout<<res<<std::endl;
}

// Your Codec object will be instantiated and called as such:
// Codec* ser = new Codec();
// Codec* deser = new Codec();
// string tree = ser->serialize(root);
// TreeNode* ans = deser->deserialize(tree);
// return ans;

int main() {
    TreeNode* root = new TreeNode(2);
    TreeNode* left = new TreeNode(1);
    TreeNode* right = new TreeNode(3);
//    TreeNode* ll = new TreeNode(1);
//    TreeNode* lr = new TreeNode(3);
//    TreeNode* rl = new TreeNode(6);
//    TreeNode* rr = new TreeNode(9);
//    [4,2,1,3,7,6,9]
    root->left = left;
    root->right = right;
//    left->left = ll;
//    left->right = lr;
//    right->left = rl;
//    right->right = rr;
    Codec().printTree(root);
    //[2,1,3]
    string s = Codec().serialize(root);
    TreeNode* ano = Codec().deserialize(s);
    vector<string> vec = Codec().printTree(ano);
    printVec(vec);
    //result
    //[2,1,3]
}

File: leetcode_0449/java/leetcode_0449.java
public class leetcode_0449 {
    public class TreeNode{
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode() {}
        TreeNode(int val) {this.val = val;}
        TreeNode(int val,TreeNode left,TreeNode right){
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }
    public class Codec {
        // Encodes a tree to a single string.
        public String serialize(TreeNode root) {
            if(root == null)
                return "#";
            StringBuilder sb = new StringBuilder();
            sb.append(root.val);
            sb.append(" ");
            sb.append(serialize(root.left));
            sb.append(" ");
            sb.append(serialize(root.right));
            return sb.toString();
        }

        // Decodes your encoded data to tree.
        public TreeNode deserialize(String data) {
            String[] splits = data.split(" ");
            index = 0;
            TreeNode root = rebuild(splits);
            return root;
        }
        private int index = 0;
        private TreeNode rebuild(String[] splits){
            if(index >= splits.length)
                return null;
            if(splits[index].equals("#")){
                index+=1;
                return null;
            }
            TreeNode node = new TreeNode(Integer.parseInt(splits[index++]));
            node.left = rebuild(splits);
            node.right = rebuild(splits);
            return node;
        }

        public void traversePreTree(StringBuilder sb,TreeNode root){
            if(root != null){
                sb.append(root.val+",");
                traversePreTree(sb,root.left);
                traversePreTree(sb,root.right);
            }
        }
        public void printTree(TreeNode root){
            StringBuilder res = new StringBuilder();
            traversePreTree(res,root);
            System.out.printf("[");
            String substring = res.substring(0, res.length() - 1);
            System.out.printf(substring);
            System.out.println("]");
        }
    }
    public static void main(String[] args) {
        leetcode_0449.Codec codec = new leetcode_0449().new Codec();
        leetcode_0449.TreeNode root = new leetcode_0449().new TreeNode(2);
        leetcode_0449.TreeNode left = new leetcode_0449().new TreeNode(1);
        leetcode_0449.TreeNode right = new leetcode_0449().new TreeNode(3);
//        leetcode_0449.TreeNode ll = new leetcode_0449().new TreeNode(1);
//        leetcode_0449.TreeNode lr = new leetcode_0449().new TreeNode(3);
//        leetcode_0449.TreeNode rl = new leetcode_0449().new TreeNode(6);
//        leetcode_0449.TreeNode rr = new leetcode_0449().new TreeNode(9);
//        left.left = ll;
//        left.right = lr;
//        right.left = rl;
//        right.right = rr;
        root.left = left;
        root.right = right;
        codec.printTree(root);
        //[2,1,3]
        String serializeStr = codec.serialize(root);
        leetcode_0449.TreeNode ano = codec.deserialize(serializeStr);
        codec.printTree(ano);
        //result
        //[2,1,3]
    }
}
